File: 4 Kyu/4 kyu - Longest Palindromic Substring (Linear).cpp
#include<iostream>
#include <vector>
#include<string>

using namespace std;

string split (string str) {
  string os;
  const int size = str.size();
  for (int i = 0; i < size; i++) {
      os += str[i];
      if (i < size - 1) os += '|';
  }
  return os;
}
string join (string str) {
  string os;
  for (auto &it : str) {
      if (it != '|')
          os += it;
  }
  return os;
}

string longest_palindrome (const string &input) {

    string str = split (input);
    const size_t size = str.size();
    size_t maxv = 0, ax = 0;
    vector<string> base;

    while (maxv < size - ax) {

        int beg = ax - 1, end = ax + 1;

        while (beg >= 0 && end < size) {
            if (str[beg] != str[end]) break;
            beg--, end++;
        }

        string sub = join (str.substr (beg + 1, end - beg - 1));
        if (sub.size() > maxv) base.push_back(sub);
        maxv = max (sub.size(), maxv);

        ax++;
    }

    for (const auto &it : base)
        if (it.size() == maxv) return it;

    return "";
}

int main () {

    string str = "ababbab";
    str = split (str);
    const size_t size = str.size();
    size_t maxv = 0, ax = 0;
    vector<string> base;

    while (maxv < size - ax) {

        int beg = ax - 1, end = ax + 1;

        while (beg >= 0 && end < size) {
            if (str[beg] != str[end]) break;
            beg--, end++;
        }

        string sub = join (str.substr (beg + 1, end - beg - 1));
        cout << sub << "\n";
        if (sub.size() > maxv) base.push_back(sub);
        maxv = max (sub.size(), maxv);

        ax++;
    }


}

File: 4 Kyu/4 kyu - Longest Palindromic Substring (Linear).rs
fn split (src: &str) -> Vec<char> {

    let mut base:Vec<char> = Vec::new();

    for it in src.chars() {
        base.push(it);
        base.push('|');
    }
    base.pop();

    base
}
fn longest_palindrome(s: &str) -> String {

    let s:Vec<char> = split (s);
    let size:usize = s.len();

    let mut maxv:usize = 0;
    let mut ax:usize = 0;
    let mut base:Vec<String> = Vec::new();

    while maxv < size - ax {
        let mut beg:i32 = ax as i32;
        let mut end:i32 = ax as i32;

        while beg >= 0 && end < size as i32 {
            if s[beg as usize] != s[end as usize] { break }
            beg -= 1;
            end += 1;
        }

        let sub = &s[(beg + 1) as usize .. (end) as usize];

        let sub:String = sub.iter().filter(|&&x| x != '|').collect();
        if sub.len() > maxv { base.push (format! ("{sub}")) }
        maxv = std::cmp::max (maxv, sub.len());

        ax += 1;
    }

    for tok in base {
        if tok.len() == maxv { return tok; }
    }

    "".to_string()
}
